### PR TITLE
Regionprops3d

### PR DESCRIPTION
**Pull request by emmanuelle**
This PR is a proposition to extend ``regionprops`` 3-D images (following #1489). Not all features of ``regionprops`` are implemented (yet), and a ``NotImplementedError`` is raised if a property not implemented for 3-D images is called on a 3-D image.

### CODE REVIEW

**Comment by stefanv** [comment]
I like it.  Persistent subterfuge.

**Comment by emmanuelle** [comment]
I just rebased on master. Do you think this PR could be merged as it is?

File: skimage/measure/_regionprops.py [hunk 1]
```diff
@@ -179,6 +195,8 @@ def image(self):
 
     @_cached_property
     def inertia_tensor(self):
+        if self._ndim > 2:
+            raise NotImplementedError('not implemented for 3D images')
```
Comment (ahojnnes) [inline skimage/measure/_regionprops.py]: This code is repeated in all functions and should be extracted into a separate method in my opinion.

Comment (ahojnnes) [inline skimage/measure/_regionprops.py]: something like, self.assert2d()

File: skimage/measure/_regionprops.py [hunk 2]
```diff
@@ -113,38 +114,47 @@ def __init__(self, slice, label, label_image, intensity_image,
         self._label_image = label_image
         self._intensity_image = intensity_image
         self._cache_active = cache_active
+        self._ndim = label_image.ndim
 
     @property
     def area(self):
-        return self.moments[0, 0]
+        return self.image.sum()
```
Comment (ahojnnes) [inline skimage/measure/_regionprops.py]: Much better.

Comment (ahojnnes) [inline skimage/measure/_regionprops.py]: Since moments is not used any more, this should become a `_cached_property`.

**Comment by jni** [comment]
@ahojnnes @stefanv can you guys review this again? It's ready to merge, imho. (But since I contributed commits I won't push the button.)

**Comment by ahojnnes** [comment]
The only thing I am skeptical about is the performance impact of the additional @only2d decorator. Other than that it looks good to me.

**Comment by emmanuelle** [comment]
@ahojnnes I ran a few tests trying to detect a possible performance difference due to @only2d, and couldn't detect anything significant. I think this PR is good to go now, can somebody review it quickly once again (it's been a long time since the last discussions...)?

@jni I had to remove the ``@property`` decorator because it made ``_install_properties_doc`` crash, what for was this decorator needed?

**Comment by jni** [comment]
@emmanuelle I thought `@property` is the reason something like this works?

```python
props = measure.regionprops(labels)
conv0 = props[0].convex_hull_area  # no function call needed
```

Am I wrong?

**Comment by emmanuelle** [comment]
@jni thanks! So we definitely need this feature. I'll figure out what was going on with tests, so please don't merge the PR for now!

**Comment by emmanuelle** [comment]
@jni : the @property decorator is not needed explicitely any more, because `_install_properties_doc` is doing the magics for all functions. At first I didn't understand what this function was meant for, so I opened #1851 to ask for more documentation. I also checked manually that the expected behaviour is still the same.

Therefore, I think this PR is ready now!

**Comment by jni** [comment]
@emmanuelle thanks for checking! Yes, I was a bit surprised that the tests would pass if the decorator was needed for this behaviour! =) I'll review today. Thank you!

File: skimage/measure/_regionprops.py [hunk 3]
```diff
@@ -476,8 +517,8 @@ def regionprops(label_image, intensity_image=None, cache=True):
 
     label_image = np.squeeze(label_image)
 
-    if label_image.ndim != 2:
-        raise TypeError('Only 2-D images supported.')
+    if label_image.ndim not in (2, 3):
+            raise TypeError('Only 2-D and 3-D images supported.')
```
Comment (jni) [inline skimage/measure/_regionprops.py]: Errant indent.

File: skimage/measure/_regionprops.py [hunk 4]
```diff
@@ -73,6 +74,16 @@ def wrapper(obj):
     return wrapper
 
 
+def only2d(method):
+    def func2d(self, *args, **kwargs):
+        if self._ndim > 2:
+            raise NotImplementedError('Property %s is not implemented for '
+                                      '3D images' % method.__name__)
+        return method(self, *args, **kwargs)
+    func2d.__name__ = method.__name__
```
Comment (jni) [inline skimage/measure/_regionprops.py]: @emmanuelle would this be better implemented with `@functools.wraps`?

File: skimage/measure/_regionprops.py [hunk 5]
```diff
@@ -134,21 +156,24 @@ def euler_number(self):
         return -num + 1
 
     def extent(self):
-        rows, cols = self.image.shape
-        return self.moments[0, 0] / (rows * cols)
+        return float(self.area) / self.image.size
```
Comment (jni) [inline skimage/measure/_regionprops.py]: I would prefer here that you use `from __future__ import division` (at top) and then just write `self.area / self.image.size`. This makes the code more Python 3-like, which is what we should aim for in general imho.

File: skimage/measure/_regionprops.py [hunk 6]
```diff
@@ -134,21 +156,24 @@ def euler_number(self):
...
     def filled_area(self):
         return np.sum(self.filled_image)
 
     @_cached
     def filled_image(self):
-        return ndi.binary_fill_holes(self.image, STREL_8)
+        if self._ndim == 2:
+            return ndi.binary_fill_holes(self.image, STREL_8)
+        else:
+            return ndi.binary_fill_holes(self.image, STREL_26_3D)
```
Comment (jni) [inline skimage/measure/_regionprops.py]: This is very nitpicky and you can take it or leave it, but while I'm at it:

```python
structure = STREL_8 if self._ndim == 2 else STREL_26_3D
return ndi.binary_fill_holes(self.image, structure)
```

**Comment by emmanuelle** [comment]
@jni there you go! Thanks a lot for the review!

**Pull request merged**